### PR TITLE
(PUP-5908) Add MsgPack serialization

### PR DESCRIPTION
**Pull request by thallgren**
This commit adds a serialization reader/writer pair that is based on
MsgPack.

### CODE REVIEW

**Comment by thallgren** [comment]
This changes the Gem requirements (msgpack becomes required). The ramifications of that needs to be ironed out before this PR is merged.

**Comment by hlindberg** [comment]
second commit message is not in imperative form

File: lib/puppet/pops/serialization/abstract_writer.rb [hunk 1]
```diff
@@ -178,12 +191,16 @@ def register_types
...
 
-    register_type(Extension::SENSITIVE, Types::PSensitiveType::Sensitive) do |o|
-      build_payload { |ep| ep.write(o.unwrap) }
-    end
-
-    register_type(Extension::BINARY, Types::PBinaryType::Binary) do |o|
-      build_payload { |ep| ep.write(o.to_s) }
+    if supports_binary?
+      register_type(Extension::BINARY, Types::PBinaryType::Binary) do |o|
+        # The Ruby MessagePack implementation have special treatment for "ASCII-8BIT" strings. They
```
Comment (hlindberg) [inline lib/puppet/pops/serialization/abstract_writer.rb]: "implementation has"

File: lib/puppet/pops/serialization/extension.rb [hunk 2]
```diff
@@ -26,8 +27,8 @@ module Extension
   TIMESPAN = 0x34
   VERSION = 0x35
   VERSION_RANGE = 0x36
-  SENSITIVE = 0x37
-  BINARY = 0x38
+  BINARY = 0x37
+  BASE64 = 0x38
```
Comment (hlindberg) [inline lib/puppet/pops/serialization/extension.rb]: Why do we have a BASE64 serialization type? Isn't that a choice per serializer how to treat a Binary ?

Comment (thallgren) [inline lib/puppet/pops/serialization/extension.rb]: It is, and a binary capable serializer (e.g. msgpack) will use BINARY whereas a non-binary serializer (e.g. json) will use BASE64 when writing to a stream. But consider what happens if one version of a serializer isn't capable whereas of writing binary content but another version is. The latter would then need to differentiate between the two when reading. Using two separate extension points (both for the Binary) makes this possible.

**Comment by melissa** [comment]
@thallgren @hlindberg do we still need this now that rich data is a thing? Can this PR be closed, or should we keep it to be reworked?

**Comment by hlindberg** [comment]
IMO, this should be reworked - Pcore over MsgPack is an improvement to both "human readable JSON" and "human readable JSON in MsgPack format" - (both actually being rich data).
The improvements come from Pcore serialization being able to tabulate data and thereby reducing size on the wire as well as memory needed at deserialized end.

**Comment by melissa** [comment]
In that case, should we close this PR? Or would it be best to leave it open?

**Comment by thallgren** [comment]
I would like this PR to be left open.

**Comment by melissa** [comment]
Great! Thank you for the update @hlindberg and @thallgren !!

**Comment by joshcooper** [comment]
> The improvements come from Pcore serialization being able to tabulate data and thereby reducing size on the wire

By default the puppetserver catalog endpoint applys `gzip` compression (true for most/all of the endpoints), where compression is handled in native code. Have you benchmarked "human readable JSON over msgpack and then gzip compressed" compared to "pcore tabulation over msgpack and then gzip compressed", where pcore tabulation occurs in ruby, and I'd assume is much slower?

> as well as memory needed at deserialized end.

Do you have any numbers for this?

Aside, several conflicts need to be resolved.

**Comment by joshcooper** [comment]
@thallgren I noticed this hasn't seen any activity in a long time, do you think you'll be able to work on it in the near future?

**Comment by joshcooper** [comment]
Since this hasn't seen any activity in over a year, I'm going to go ahead and close this pull request for the time being. Please re-open this pull request once the next actions are addressed (conflicts and gem dependencies issue are resolved).

**Pull request closed without merging**